icu4j/main/classes/core/src/com/ibm/icu/message2/Mf2DataModelFormatter.java

```diff
@@ -0,0 +1,277 @@
+
+        standardFunctionRegistry = Mf2FunctionRegistry.builder()
+                // Date/time formatting
+                .setFormatter("datetime", new DateTimeFormatterFactory())
+                .setDefaultFormatterNameForType(Date.class, "datetime")
+                .setDefaultFormatterNameForType(Calendar.class, "datetime")
+
+                // Number formatting
+                .setFormatter("number", new NumberFormatterFactory())
+                .setDefaultFormatterNameForType(Integer.class, "number")
+                .setDefaultFormatterNameForType(Double.class, "number")
+                .setDefaultFormatterNameForType(Number.class, "number")
+                .setDefaultFormatterNameForType(CurrencyAmount.class, "number")
+
+                // Format that returns "to string"
+                .setFormatter("identity", new IdentityFormatterFactory())
+                .setDefaultFormatterNameForType(String.class, "identity")
+                .setDefaultFormatterNameForType(CharSequence.class, "identity")
+
+                // Register the standard selectors
+                .setSelector("plural", new PluralSelectorFactory("cardinal"))
+                .setSelector("selectordinal", new PluralSelectorFactory("ordinal"))
+                .setSelector("select", new TextSelectorFactory())
+                .setSelector("gender", new TextSelectorFactory())
+
+                .build();
+    }
+
+    private static Map<String, Object> mf2OptToFixedOptions(Map<String, Value> options) {
+        Map<String, Object> result = new HashMap<>();
+        for (Entry<String, Value> option : options.entrySet()) {
+            Value value = option.getValue();
+            if (value.isLiteral()) {
+                result.put(option.getKey(), value.getLiteral());
+            }
+        }
+        return result;
+    }
+
+    private Map<String, Object> mf2OptToVariableOptions(Map<String, Value> options, Map<String, Object> arguments) {
+        Map<String, Object> result = new HashMap<>();
+        for (Entry<String, Value> option : options.entrySet()) {
+            Value value = option.getValue();
+            if (value.isVariable()) {
+                result.put(option.getKey(), variableToObjectEx(value, arguments));
+            }
+        }
+        return result;
+    }
+
+    FormatterFactory getFormattingFunctionFactoryByName(Object toFormat, String functionName) {
+        // Get a function name from the type of the object to format
+        if (functionName == null || functionName.isEmpty()) {
+            if (toFormat == null) {
+                // The object to format is null, and no function provided.
+                return null;
+            }
+            Class<?> clazz = toFormat.getClass();
+            functionName = standardFunctionRegistry.getDefaultFormatterNameForType(clazz);
+            if (functionName == null) {
+                functionName = customFunctionRegistry.getDefaultFormatterNameForType(clazz);
+            }
```

Comment:
Q1: in is not specified who wins. I can open a ticket.

I think one can argue both ways...
I decided to have the standard win on the idea that the standard is the standard and one should not mess with it.
If you want a different behavior, make a custom.

Q2. I think we need both.
For example you format `"... {$someone :person} ... was born on {$dob :date skeleton=yMMMd}..."`

But the I also wanted to talk with you (after the release pressure is gone) about being able to separate the standard formatters for backport (for example implement them on top of Android APIs).

But yes, the current implementation ties them together, because it is in ICU anyway.

Reply:
Discussed.

Reply:
Part of the discussion: There should not be a name collision, because a custom registry should not use standard function names. The MFWG has discussed a naming convention, such as custom function names containing dots (e.g., "com.google.somegooglefunction").

Reply:
Created issue: https://github.com/unicode-org/message-format-wg/issues/298